<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.335">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Bash for Bioinformatics - 6&nbsp; Batch Processing on the Cloud</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./06-JSON.html" rel="next">
<link href="./04-doing-work-with-dx-run.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
<script async="" src="https://www.googletagmanager.com/gtag/js?id=G-QM3D3CT5J5"></script>

<script type="text/javascript">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-QM3D3CT5J5', { 'anonymize_ip': true});
</script>
<script type="application/json" class="js-hypothesis-config">
{
  "theme": "clean",
  "openSidebar": false
}
</script>
<script async="" src="https://hypothes.is/embed.js"></script>
<script src="site_libs/quarto-diagram/mermaid.min.js"></script>
<script src="site_libs/quarto-diagram/mermaid-init.js"></script>
<link href="site_libs/quarto-diagram/mermaid.css" rel="stylesheet">


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Batch Processing on the Cloud</span></h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Bash for Bioinformatics</a> 
    </div>
      </div>
      <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./01-setup-course.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Setup for the Course / dx-toolkit basics</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./02-scripting-basics.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Shell Scripting Basics</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./03-cloud-computing-basics.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Cloud Computing Basics</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./04-doing-work-with-dx-run.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Working with files using <code>dx run</code></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./05-batch-processing.html" class="sidebar-item-text sidebar-link active"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Batch Processing on the Cloud</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./06-JSON.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Working with JSON on the DNAnexus Platform</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./07-containers.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Containers and Reproducibility</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./appendix.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Appendix</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">References</a>
  </div>
</li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#learning-objectives" id="toc-learning-objectives" class="nav-link active" data-scroll-target="#learning-objectives"><span class="toc-section-number">6.1</span>  Learning Objectives</a></li>
  <li><a href="#two-ways-of-batching" id="toc-two-ways-of-batching" class="nav-link" data-scroll-target="#two-ways-of-batching"><span class="toc-section-number">6.2</span>  Two Ways of Batching</a></li>
  <li><a href="#sec-dx-find" id="toc-sec-dx-find" class="nav-link" data-scroll-target="#sec-dx-find"><span class="toc-section-number">6.3</span>  Finding files using <code>dx find data</code></a></li>
  <li><a href="#helpful-dx-find-data-examples" id="toc-helpful-dx-find-data-examples" class="nav-link" data-scroll-target="#helpful-dx-find-data-examples"><span class="toc-section-number">6.4</span>  Helpful <code>dx find data</code> examples</a>
  <ul class="collapse">
  <li><a href="#find-all-.bam-files-in-a-project" id="toc-find-all-.bam-files-in-a-project" class="nav-link" data-scroll-target="#find-all-.bam-files-in-a-project"><span class="toc-section-number">6.4.1</span>  Find all *.bam files in a project</a></li>
  <li><a href="#searching-within-a-folder" id="toc-searching-within-a-folder" class="nav-link" data-scroll-target="#searching-within-a-folder"><span class="toc-section-number">6.4.2</span>  Searching within a folder</a></li>
  <li><a href="#find-all-files-with-a-field-id" id="toc-find-all-files-with-a-field-id" class="nav-link" data-scroll-target="#find-all-files-with-a-field-id"><span class="toc-section-number">6.4.3</span>  Find all files with a field id</a></li>
  <li><a href="#find-all-files-that-are-of-class-file" id="toc-find-all-files-that-are-of-class-file" class="nav-link" data-scroll-target="#find-all-files-that-are-of-class-file"><span class="toc-section-number">6.4.4</span>  Find all files that are of class <code>file</code></a></li>
  <li><a href="#in-general-think-about-leveraging-metadata" id="toc-in-general-think-about-leveraging-metadata" class="nav-link" data-scroll-target="#in-general-think-about-leveraging-metadata"><span class="toc-section-number">6.4.5</span>  In General: Think about leveraging metadata</a></li>
  </ul></li>
  <li><a href="#sec-xargs2" id="toc-sec-xargs2" class="nav-link" data-scroll-target="#sec-xargs2"><span class="toc-section-number">6.5</span>  Using <code>xargs</code> to Batch Multiple Files</a>
  <ul class="collapse">
  <li><a href="#when-batching-tag-your-jobs" id="toc-when-batching-tag-your-jobs" class="nav-link" data-scroll-target="#when-batching-tag-your-jobs"><span class="toc-section-number">6.5.1</span>  When batching, tag your jobs</a></li>
  <li><a href="#sec-terminate" id="toc-sec-terminate" class="nav-link" data-scroll-target="#sec-terminate"><span class="toc-section-number">6.5.2</span>  Using tags to <code>dx terminate</code> jobs</a></li>
  </ul></li>
  <li><a href="#sec-mult-worker" id="toc-sec-mult-worker" class="nav-link" data-scroll-target="#sec-mult-worker"><span class="toc-section-number">6.6</span>  Submitting Multiple Files to a Single Worker</a></li>
  <li><a href="#batching-multiple-inputs-dx-generate_batch_inputs" id="toc-batching-multiple-inputs-dx-generate_batch_inputs" class="nav-link" data-scroll-target="#batching-multiple-inputs-dx-generate_batch_inputs"><span class="toc-section-number">6.7</span>  Batching multiple inputs: <code>dx generate_batch_inputs</code></a>
  <ul class="collapse">
  <li><a href="#drawbacks-to-dx-generate_batch_inputsdx-run---batch-tsv" id="toc-drawbacks-to-dx-generate_batch_inputsdx-run---batch-tsv" class="nav-link" data-scroll-target="#drawbacks-to-dx-generate_batch_inputsdx-run---batch-tsv"><span class="toc-section-number">6.7.1</span>  Drawbacks to <code>dx generate_batch_inputs</code>/<code>dx run --batch-tsv</code></a></li>
  <li><a href="#for-more-information" id="toc-for-more-information" class="nav-link" data-scroll-target="#for-more-information"><span class="toc-section-number">6.7.2</span>  For More Information</a></li>
  </ul></li>
  <li><a href="#programatically-submitting-arrays-of-files-for-a-job" id="toc-programatically-submitting-arrays-of-files-for-a-job" class="nav-link" data-scroll-target="#programatically-submitting-arrays-of-files-for-a-job"><span class="toc-section-number">6.8</span>  Programatically Submitting Arrays of Files for a job</a></li>
  <li><a href="#what-you-learned-in-this-chapter" id="toc-what-you-learned-in-this-chapter" class="nav-link" data-scroll-target="#what-you-learned-in-this-chapter"><span class="toc-section-number">6.9</span>  What you learned in this chapter</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span id="sec-batch" class="quarto-section-identifier d-none d-lg-block"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Batch Processing on the Cloud</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<p>Now we’re prepared for the big one: batch processing on the DNAnexus platform. All of the shell and DNAnexus skills we’ve learned will be leveraged in this chapter.</p>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Prep for Exercises
</div>
</div>
<div class="callout-body-container callout-body">
<p>Make sure you are logged into the platform using <code>dx login</code> and that your course project is selected with <code>dx select</code>.</p>
<p>In your shell (either on your machine or in binder), make sure you’re in the <code>bash_bioinfo_scripts/batch-processing/</code> folder:</p>
<pre><code>cd batch-processing/</code></pre>
</div>
</div>
<section id="learning-objectives" class="level2" data-number="6.1">
<h2 data-number="6.1" class="anchored" data-anchor-id="learning-objectives"><span class="header-section-number">6.1</span> Learning Objectives</h2>
<ol type="1">
<li><strong>Utilize</strong> <code>dx find data</code> to find data files on the platform to batch process.</li>
<li><strong>Iterate</strong> over files using Bash scripting and <code>xargs</code> on the platform to batch process them within a DNAnexus project.</li>
<li><strong>Leverage</strong> dxFUSE to simplify your bash scripts</li>
<li><strong>Utilize</strong> <code>dx generate-batch-inputs</code>/<code>dx run --batch-tsv</code> to batch process files</li>
<li><strong>Utilize</strong> Python to batch process multiple files per worker.</li>
</ol>
</section>
<section id="two-ways-of-batching" class="level2" data-number="6.2">
<h2 data-number="6.2" class="anchored" data-anchor-id="two-ways-of-batching"><span class="header-section-number">6.2</span> Two Ways of Batching</h2>
<div id="fig-batch1" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<div class="cell">
<div class="cell-output-display">
<div>
<p>
</p><pre class="mermaid mermaid-js" data-tooltip-selector="#mermaid-tooltip-1">graph LR;
  A[List files &lt;/br&gt; using `dx data`] --&gt; F{"|"}
  F --&gt; E[`xargs` sh -c]
  E --&gt; B[`dx run` &lt;/br&gt; on file1];
  E --&gt; C[`dx run` &lt;/br&gt; on file2];
  E --&gt; D[`dx run` &lt;/br&gt; on file3];
</pre>
<div id="mermaid-tooltip-1" class="mermaidTooltip">

</div>
<p></p>
</div>
</div>
</div>
<p></p><figcaption class="figure-caption">Figure&nbsp;6.1: Batch method 1. We list files and then pipe them into <code>xargs</code>, which generates individual dx-run statements.</figcaption><p></p>
</figure>
</div>
<div id="fig-batch2" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<div class="cell">
<div class="cell-output-display">
<div>
<p>
</p><pre class="mermaid mermaid-js" data-tooltip-selector="#mermaid-tooltip-2">graph LR;
  A[Submit array &lt;/br&gt; of files &lt;/br&gt; in `dx run`] --&gt; B[Loop over array &lt;/br&gt; of files &lt;/br&gt; in worker];
</pre>
<div id="mermaid-tooltip-2" class="mermaidTooltip">

</div>
<p></p>
</div>
</div>
</div>
<p></p><figcaption class="figure-caption">Figure&nbsp;6.2: Batch method 2. We first get our files onto the worker through a single dx run command, and then use <code>xargs</code> on the worker to cycle through them.</figcaption><p></p>
</figure>
</div>
<p>We actually have two methods of batching jobs using Swiss Army Knife:</p>
<ol type="1">
<li>Use <code>xargs</code> on our home system to run <code>dx run</code> statements for each file (<a href="#fig-batch1">Figure&nbsp;<span>6.1</span></a>).</li>
<li>Submit an array of files as an input to Swiss Army Knife. Then process each file using the <code>icmd</code> input (<a href="#fig-batch2">Figure&nbsp;<span>6.2</span></a>)</li>
</ol>
<p>Both of these methods can potentially be useful.</p>
</section>
<section id="sec-dx-find" class="level2" data-number="6.3">
<h2 data-number="6.3" class="anchored" data-anchor-id="sec-dx-find"><span class="header-section-number">6.3</span> Finding files using <code>dx find data</code></h2>
<p><code>dx find data</code> is a command that is extremely helpful on the DNAnexus platform. Based on metadata and folder paths, <code>dx find data</code> will return a list of files that meet the criteria.</p>
<p><code>dx find data</code> lets you search on the following types of metadata:</p>
<ul>
<li>tags <code>--tag</code></li>
<li>properties <code>--property</code></li>
<li>name <code>--name</code></li>
<li>type <code>--type</code></li>
</ul>
<p>It can output in a number of different formats. Including:</p>
<ul>
<li><code>--brief</code> - return only the file-ids</li>
<li><code>--json</code> - return file information in JSON format</li>
<li><code>--verbose</code> - this is the default setting</li>
<li><code>--delimited</code> - return as a delimited text file</li>
</ul>
<p>Of all of these, <code>--brief</code> and <code>--json</code> are the most useful for automation. <code>--delimited</code> is also helpful, but there is also a utility called <code>dx generate-batch-inputs</code> that will let us specify multiple inputs to process line by line.</p>
</section>
<section id="helpful-dx-find-data-examples" class="level2" data-number="6.4">
<h2 data-number="6.4" class="anchored" data-anchor-id="helpful-dx-find-data-examples"><span class="header-section-number">6.4</span> Helpful <code>dx find data</code> examples</h2>
<p>As we’re starting off in our batch processing journey, I wanted to provide some helpful recipes for selecting files.</p>
<section id="find-all-.bam-files-in-a-project" class="level3" data-number="6.4.1">
<h3 data-number="6.4.1" class="anchored" data-anchor-id="find-all-.bam-files-in-a-project"><span class="header-section-number">6.4.1</span> Find all *.bam files in a project</h3>
<p>You can use wildcard characters with the <code>--name</code> flag. Here, we’re looking for anything with the suffix “*.bam”.</p>
<pre class="{bash}"><code>#| eval: false
#| filename: batch-processing/dx-find-data-name.sh
dx find data --name "*.bam" --brief</code></pre>
</section>
<section id="searching-within-a-folder" class="level3" data-number="6.4.2">
<h3 data-number="6.4.2" class="anchored" data-anchor-id="searching-within-a-folder"><span class="header-section-number">6.4.2</span> Searching within a folder</h3>
<p>You can add the <code>--path</code> command to search in a specific folder.</p>
<pre class="{bash}"><code>#| eval: false
#| filename: batch-processing/dx-find-path.sh

dx find data --name "*bam" --path "data/"</code></pre>
</section>
<section id="find-all-files-with-a-field-id" class="level3" data-number="6.4.3">
<h3 data-number="6.4.3" class="anchored" data-anchor-id="find-all-files-with-a-field-id"><span class="header-section-number">6.4.3</span> Find all files with a field id</h3>
<p>Take advantage of metadata associated with files when you can. If you are on UKB RAP, one of the most helpful properties to search is <code>field_id</code>.</p>
<p>Note: be careful with this one, especially if you are working on UK Biobank RAP. You don’t want to return 500,000 file ids. I would concentrate on the field ids that are aggregated on the population level, such as the pVCF files.</p>
<pre class="{bash}"><code>#| eval: false
#| filename: batch-processing/dx-find-data-field.sh

dx find data --property field_id=23148 --brief</code></pre>
</section>
<section id="find-all-files-that-are-of-class-file" class="level3" data-number="6.4.4">
<h3 data-number="6.4.4" class="anchored" data-anchor-id="find-all-files-that-are-of-class-file"><span class="header-section-number">6.4.4</span> Find all files that are of class <code>file</code></h3>
<p>There are a number of different object classes on the platform, such as <code>file</code> or <code>applet</code></p>
<p>Search for all files in your project that have a <code>file</code> class.</p>
<pre class="{bash}"><code>#| eval: false
#| filename: batch-processing/dx-find-data-class.sh

dx find data --class file --brief</code></pre>
</section>
<section id="in-general-think-about-leveraging-metadata" class="level3" data-number="6.4.5">
<h3 data-number="6.4.5" class="anchored" data-anchor-id="in-general-think-about-leveraging-metadata"><span class="header-section-number">6.4.5</span> In General: Think about leveraging metadata</h3>
<p>In general, think about leveraging metadata that is attached to your files.</p>
<p>For example, for the UKB Research Analysis Platform, data files in the <code>Bulk/</code> folder in your project have multiple properties: <code>field_id</code> (the data field as specified by UK Biobank) and <code>eid</code>.</p>
</section>
</section>
<section id="sec-xargs2" class="level2" data-number="6.5">
<h2 data-number="6.5" class="anchored" data-anchor-id="sec-xargs2"><span class="header-section-number">6.5</span> Using <code>xargs</code> to Batch Multiple Files</h2>
<p>Ok, now we have a list of files from <code>dx find data</code> that meet our criteria. How can we process them one by one?</p>
<p>Remember our discussion of <code>xargs</code>? (<a href="02-scripting-basics.html#sec-xargs"><span>Section&nbsp;3.5</span></a>) This is where <code>xargs</code> shines, when you provide it a list of files.</p>
<p>Remember, a really useful pattern for <code>xargs</code> is using it for variable expansion and starting a subshell to process individual files.</p>
<pre class="{bash}"><code>#| eval: false
#| filename: batch-processing/dx-find-xargs.sh
dx find data --name "*.bam" --brief | \
  xargs -I % sh -c "dx run app-swiss-army-knife -y -iin="%" \
  -icmd='samtools view -c \${in_name} &gt; \${in_prefix-counts.txt}' \
  --tag samjob --destination results/' </code></pre>
<p>The key piece of code we’re doing the variable expansion in is here:</p>
<pre class="{bash}"><code>#| eval: false
sh -c 'dx run app-swiss-army-knife -iin="%" \ 
  -icmd="samtools view -c \${in_name}\ &gt; \${in_prefix}-counts.txt" \
  --tag samjob --destination results/'</code></pre>
<p>We’re using <code>sh -c</code> to run a script as a <em>subshell</em> to execute the <code>dx run</code> statement.</p>
<p>Note that we’re specifying the helper variables here with a <code>\</code>:</p>
<p><code>\${in_name}</code></p>
<p>This escaping (<code>\$</code>) of the dollar sign is to prevent the variable expansion from happening in the top-level shell - the helper variable names need to be passed in to the subshell which needs to pass it onto the worker. Figuring this out took time and made my brain hurt.</p>
<p>This escaping is only necessary because we’re using <code>xargs</code> and passing our <code>-icmd</code> input into the worker. For the most part, you won’t need to escape the <code>$</code>. This is also a reason to write shell scripts that run on the worker.</p>
<p>When we run this command, we get the following screen output:</p>
<pre><code>Using input JSON:
{
    "cmd": "samtools view -c $in_name &gt; $in_prefix-counts.txt",
    "in": [
        {
            "$dnanexus_link": {
                "project": "project-GGyyqvj0yp6B82ZZ9y23Zf6q",
                "id": "file-BZ9YGpj0x05xKxZ42QPqZkJY"
            }
        }
    ]
}

Calling app-GFxJgVj9Q0qQFykQ8X27768Y with output destination
  project-GGyyqvj0yp6B82ZZ9y23Zf6q:/results

Job ID: job-GJ2xVZ80yp62X5Z51qp191Y8

[more job info]</code></pre>
<p>if we do a <code>dx find jobs</code>, we’ll see our jobs listed. Hopefully they are running:</p>
<pre><code>dx find jobs --tag samjob
* Swiss Army Knife (swiss-army-knife:main) (running) job-GJ2xVf00yp62kx9Z8VK10vpQ
  tladeras 2022-10-11 13:57:59 (runtime 0:01:49)
* Swiss Army Knife (swiss-army-knife:main) (running) job-GJ2xVb80yp6KjQpxFJJBzv5k
  tladeras 2022-10-11 13:57:57 (runtime 0:00:52)
* Swiss Army Knife (swiss-army-knife:main) (runnable) job-GJ2xVZj0yp6FFFXG11j6YJ9V
  tladeras 2022-10-11 13:57:55 (runtime 0:01:15)
* Swiss Army Knife (swiss-army-knife:main) (runnable) job-GJ2xVZ80yp62X5Z51qp191Y8
  tladeras 2022-10-11 13:57:53 (runtime 0:00:56)</code></pre>
<section id="when-batching-tag-your-jobs" class="level3" data-number="6.5.1">
<h3 data-number="6.5.1" class="anchored" data-anchor-id="when-batching-tag-your-jobs"><span class="header-section-number">6.5.1</span> When batching, tag your jobs</h3>
<p>It is critical that you tag your jobs in your <code>dx run</code> code with the <code>--tag</code> argument.</p>
<p>Why? You will at some point start up a bunch of batch jobs that might have some settings/parameters that were set wrong. That’s when you need the tag.</p>
<pre class="{bash}"><code>#| eval: false

dx find jobs --tag "samjob"</code></pre>
</section>
<section id="sec-terminate" class="level3" data-number="6.5.2">
<h3 data-number="6.5.2" class="anchored" data-anchor-id="sec-terminate"><span class="header-section-number">6.5.2</span> Using tags to <code>dx terminate</code> jobs</h3>
<p><code>dx terminate &lt;jobid&gt;</code> will terminate a running job with that job id. It doesn’t take a tag as input.</p>
<p>But again, <code>xargs</code> to the rescue. We can find our job ids with the tag <code>samjob</code> using <code>dx find jobs</code> and then pipe the <code>--brief</code> output into <code>xargs</code> to terminate each job id.</p>
<pre class="{bash}"><code>#| eval: false
dx find jobs --tag samjob --brief | xargs -I% sh -c "dx terminate %"</code></pre>
</section>
</section>
<section id="sec-mult-worker" class="level2" data-number="6.6">
<h2 data-number="6.6" class="anchored" data-anchor-id="sec-mult-worker"><span class="header-section-number">6.6</span> Submitting Multiple Files to a Single Worker</h2>
<p>We talked about another method to batch process files on a worker (<a href="#fig-batch2">Figure&nbsp;<span>6.2</span></a>). We can submit an array of files to a worker, and then process them one at a time on the worker.</p>
<p>The key is that we’re running <code>xargs</code> on the worker, not on our own machine to process each file.</p>
<pre class="{bash}"><code>#| eval: false
#| filename: batch-processing/batch-on-worker.sh
cmd_to_run="ls *.vcf.gz | xargs -I% sh -c 'bcftools stats % &gt; \$(basename %).stats.txt'"

dx run swiss-army-knife \
  -iin="data/chr1.vcf.gz" \
  -iin="data/chr2.vcf.gz" \
  -iin="data/chr3.vcf.gz" \
  -icmd=${cmd_to_run}</code></pre>
<p>In the variable <code>$cmd_to_run</code>, we’re putting a command that we’ll run on the worker. That command is:</p>
<pre class="{bash}"><code>#| eval: false
ls *.vcf.gz | xargs -I% sh -c "bcftools stats % &gt; \$(basename %).stats.txt</code></pre>
<p>We submitted an array of files in our <code>dx run</code> statement. So now they are transferred into our working directory on the worker. So we can list the files using <code>ls *.vcf.gz</code> and pipe that list into <code>xargs</code>.</p>
<p>Note that we lose the ability to use helper variables in our script when we process a list of files on the worker. So here we have to use <code>\$(basename %)</code>, because we use <code>()</code> to expand a variable in a subshell, and we escape the <code>$</code> here so that bash will execute the variable expansion on the worker.</p>
<p>Again, this is possible, but it may be easier to have a separate script that contains our commands, transfer that as an input to Swiss Army Knife, and run that script by specifying <code>bash myscript.sh</code> in our command.</p>
</section>
<section id="batching-multiple-inputs-dx-generate_batch_inputs" class="level2" data-number="6.7">
<h2 data-number="6.7" class="anchored" data-anchor-id="batching-multiple-inputs-dx-generate_batch_inputs"><span class="header-section-number">6.7</span> Batching multiple inputs: <code>dx generate_batch_inputs</code></h2>
<p>What if you have multiple inputs that you need to batch with? This is where the <a href="https://documentation.dnanexus.com/user/running-apps-and-workflows/running-batch-jobs"><code>dx generate_batch_inputs</code></a> comes in.</p>
<p>For each input for an app, we can specify it using wildcard characters.</p>
<pre class="{bash}"><code># | eval: false

dx generate_batch_inputs \
  --path "data/"\
  -iin="(.*)\.bam$"</code></pre>
<p>Here we’re specifying a single input <code>in</code>, and we’ve supplied a wildcard search. It’s going to look in <code>data/</code> for this particular pattern (we’re looking for bam files).</p>
<p>If we do this, we’ll get the following response:</p>
<pre><code>Found 4 valid batch IDs matching desired pattern.
Created batch file dx_batch.0000.tsv</code></pre>
<p>So, there is 1 <code>.tsv</code> file that was generated by <code>dx generate_batch_inputs</code> on our machine.</p>
<p>If we have many more input files, say 3000 files, it would generate 3 <code>.tsv</code> files. Each of these <code>.tsv</code> files contains about 1000 files per line. We can run these individual jobs with:</p>
<pre class="{bash}"><code>#| eval: false

dx run swiss-army-knife --batch-tsv dx_batch.0000.tsv \
   -icmd='samtools stats ${in_name} &gt; ${in_prefix}.stats.txt ' \
   --destination "/Results/" \
   --detach --allow-ssh \
   --tag bigjob</code></pre>
<p>This will generate 4 jobs from the <code>dx_batch.0000</code> file to process the individual files. Each <code>tsv</code> file will generate up to 1000 jobs.</p>
<section id="drawbacks-to-dx-generate_batch_inputsdx-run---batch-tsv" class="level3" data-number="6.7.1">
<h3 data-number="6.7.1" class="anchored" data-anchor-id="drawbacks-to-dx-generate_batch_inputsdx-run---batch-tsv"><span class="header-section-number">6.7.1</span> Drawbacks to <code>dx generate_batch_inputs</code>/<code>dx run --batch-tsv</code></h3>
<p>The largest drawback to using <code>dx generate_batch_inputs</code> is that each column must correspond to an individual input name - you can’t submit an array of files to a job this way.</p>
</section>
<section id="for-more-information" class="level3" data-number="6.7.2">
<h3 data-number="6.7.2" class="anchored" data-anchor-id="for-more-information"><span class="header-section-number">6.7.2</span> For More Information</h3>
<p>The Batch Jobs documentation page has some good code examples for <code>dx generate_batch_inputs</code> here: <a href="https://documentation.dnanexus.com/user/running-apps-and-workflows/running-batch-jobs/" class="uri">https://documentation.dnanexus.com/user/running-apps-and-workflows/running-batch-jobs/</a></p>
</section>
</section>
<section id="programatically-submitting-arrays-of-files-for-a-job" class="level2" data-number="6.8">
<h2 data-number="6.8" class="anchored" data-anchor-id="programatically-submitting-arrays-of-files-for-a-job"><span class="header-section-number">6.8</span> Programatically Submitting Arrays of Files for a job</h2>
<p>You can also use Python to build <code>dx run</code> statements, which is especially helpful when you want to submit arrays of 100+ files to a worker.</p>
<p>See <a href="https://dnanexus.gitbook.io/uk-biobank-rap/science-corner/guide-to-analyzing-large-sample-sets" class="uri">https://dnanexus.gitbook.io/uk-biobank-rap/science-corner/guide-to-analyzing-large-sample-sets</a> for more info.</p>
</section>
<section id="what-you-learned-in-this-chapter" class="level2" data-number="6.9">
<h2 data-number="6.9" class="anchored" data-anchor-id="what-you-learned-in-this-chapter"><span class="header-section-number">6.9</span> What you learned in this chapter</h2>
<p>This was a big chapter, and built on everything you’ve learned in the previous chapters.</p>
<p>We put together the output of <code>dx find data --brief</code> (<a href="#sec-dx-find"><span>Section&nbsp;6.3</span></a>) with a pipe (<code>|</code>), and used <code>xargs</code> (<a href="#sec-xargs2"><span>Section&nbsp;6.5</span></a>) to spawn jobs per set of files.</p>
<p>Another way to process files is to upload them onto a worker and process them (<a href="#sec-mult-worker"><span>Section&nbsp;6.6</span></a>).</p>
<p>We also learned of alternative approaches using <code>dx generate_batch_inputs</code>/<code>dx run --batch-tsv</code> and using Python to build the <code>dx run</code> statements.</p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./04-doing-work-with-dx-run.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Working with files using <code>dx run</code></span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./06-JSON.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Working with JSON on the DNAnexus Platform</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>